lambda/value_rec_compiler.ml

```diff
@@ -458,6 +573,65 @@ and rebuild_arms :
     | Reachable _, Reachable _ ->
       Misc.fatal_error "letrec: multiple functions"
 
+(** {1. Compilation} *)
+
+(** The bindings are split into three categories.
+    Static bindings are the ones that we can pre-allocate and backpatch later.
+    Function bindings are syntactic functions.
+    Dynamic bindings are non-recursive expressions.
+
+    The evaluation order is as follows:
+    - Evaluate all dynamic bindings
+    - Pre-allocate all static bindings
+    - Define all functions
+    - Backpatch all static bindings
+
+    Constants (and unreachable expressions) end up in the dynamic category,
+    because we substitute all occurrences of recursive variables in their
+    definition by a dummy expression, making them non-recursive.
+
+    This is correct because:
+    - [Value_rec_check] ensured that they never dereference the value of
+      those recursive variables
+    - their final value cannot depend on them either.
+
+    Functions that are not already in syntactic form also generate an additional
+    binding for the context. This binding fits into the static category.
+
+    {[
+      (* Input *)
+      let rec a x = b x
+      and b =
+        let tbl = Hashtbl.make 17 in
+        fun x -> ... (tbl, c, a) ...
+      and c = Some (d, default)
+      and d = Array.make 5 0
+      and default =
+        let _ = a in
+        0
+
+      (* Output *)
+        (* Dynamic bindings *)
+      let d = Array.make 5 0
+      let default =
+        let _ = *dummy_rec_value* in
+        0
+        (* Pre-allocations *)
+      let c = caml_alloc_dummy 2
+      let b_context = caml_alloc_dummy 1
+        (* Functions *)
+      let rec a x = b x
+      and b =
+        fun x -> ... (b_context.tbl, c, a) ...
+        (* Backpatching *)
+      let () =
+        caml_update_dummy c (Some (d, default));
+        caml_update_dummy b_context
+          (let tbl = Hashtbl.make 17 in
+           { tbl })
+    ]}
```

Comment:
In retrospect I think that the following would be slightly easier to read:

```
    Example input:
    {[
      let rec a x =
        (* syntactic function *)
        b x
      and b =
        (* non-syntactic function *)
        let tbl = Hashtbl.make 17 in
        fun x -> ... (tbl, c, a) ...
      and c =
        (* block *)
        Some (d, default)
      and d =
        (* 'dynamic' value (not recursive *)
        Array.make 5 0
      and default =
        (* constant, with (spurious) use
           of a recursive neighbor *)
        let _ = a in
        42
    ]}

    Example output:
    {[
      (* Dynamic bindings *)
      let d = Array.make 5 0
      let default =
        let _ = *dummy_rec_value* in
        42

      (* Pre-allocations *)
      let c = caml_alloc_dummy 2
      let b_context = caml_alloc_dummy 1

      (* Functions *)
      let rec a x = b x
      and b =
        fun x -> ... (b_context.tbl, c, a) ...

      (* Backpatching *)
      let () =
        caml_update_dummy c (Some (d, default));
        caml_update_dummy b_context
          (let tbl = Hashtbl.make 17 in
           { tbl })
    ]}
```

(I also added some comments on the input example to classify them (roughly), and changed the constant 0 into the constant 42 to make it very clear that the occurrence in the output is not a default choice to be re-declared/shadowed later.